Rawdata explorer widget

MathieuTillet opened the pull request:
**Description of work.**
This PR introduces the Raw Data Explorer widget, designed to browse raw data files in a streamlined way.

The widget appears tiled under the workspace list, and allow the user to select a directory and browse starting from there. Every time they click on a file, it is loaded and displayed in what is supposed to be the most appropriate way. Only one preview can exist at a time, so unused ones are closed as soon as they are not necessary.

By holding the control key, the user can accumulate different files, effectively summing them. Un-accumulating is not supported.

When the total memory used goes over 80% by Mantid count, unused workspaces are deleted until the memory goes down to a more reasonable value, starting by the ones that have not been interacted with for the longest time.

Testing this widget mainly consists in using it. It is supposed to support all the raw data that Mantid can load, or handle gracefully errors. In particular, since it has been developed exclusively on Ubuntu, checking that Qt behaves on other system would be good. 

Attached [RDExp.pdf](https://code.ill.fr/scientific-software/mantill/uploads/e6d2c32799c12a26be0ff059c5a53bf4/RDExp.pdf) is the requirement document, which might be useful for reference of expected behaviors and features.
<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

1. Open workbench.
2. Find the Raw data explorer widget, by default it should be placed behind the `Workspaces` widget.
3. Find some data to visualize: UnitTest/ILL data would have different types of plots from 1D plot, 2D plot, `InstrumentView`, while other instruments from outside ILL would be visualized in the `InstrumentView` widget at the moment.
4. You should  be able to navigate the tree with keyboard and each time you change a file, a new visualization should be displayed (or an informative error dialog).
5. Try to accumulate data: select a number of files while holding the `Ctrl` key. Displayed data should be added in the visualization.

Fixes #29249 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

Release note and documentation added.

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

MathieuTillet committed:
Display new raw data widget and create backbone

MathieuTillet committed:
Display files and show on click

In the rdexp widget, add the file management and a very rudimentary
display mecanism.

MathieuTillet committed:
Add an acquistion tag in facilities.xml

MathieuTillet committed:
Add a acquisition parameter to preview objects

MathieuTillet committed:
Add tests for the previews new field

MathieuTillet committed:
Add a geometry field to the instruments parameters

This is a simple tag that indicates if an instrument is 1D or 2D.

MathieuTillet committed:
Add preview declarations

Only for some ILL instruments for which all the info are already here.

MathieuTillet committed:
Manage most of the displays in the RDexp

In the raw data explorer, most of the display types are now supported.
Slice Viewer and Instrument view needs polishing to manage their
behaviour when closing.

Commit:
[IVIEW] send signal when view is closed

The view sends a signal when it is closed through the close button. This
signals is caught by the presenter to remove the instrument viewer from
the manager.

Commit:
[SVIEW] add a manager to keep open instances

Commit:
Add ILL/IN4 acquisition mode

Commit:
Add ILL/IN5 acquisition mode

Commit:
Add ILL/IN6 acquisition mode

Commit:
Add ILL/IN16B acquisition mode

Commit:
Add ILL/D11 acquisition mode

Commit:
Add ILL/D16 acquisition mode

Commit:
Add ILL/D22 acquisition mode

Commit:
Change ILL/D33 acquisition mode

Commit:
Add ILL/D2B acquisition mode

Commit:
Add ILL/D20 acquisition mode

Commit:
Add ILL/D1B acquisition mode

Commit:
Add ILL/D17 acquisition mode

Commit:
Add ILL/FIGARO acquisition mode

Commit:
Add ILL/PANTHER acquisition mode

Commit:
Add ILL/SHARP acquisition mode

Commit:
Add ILL/D7 acquisition mode

Commit:
[IVIEW] remove useless comment

Commit:
Adapt technique for ILL spectroscopy instruments

Commit:
Adapt technique for ILL backscattering instrument

Commit:
Adapt technique for ILL powder instruments

Commit:
Adapt technique for ILL polarDiff instrument

Commit:
[RDEXP] preview for spec ILL instruments

Commit:
[RDEXP] preview for SANS ILL instruments

Commit:
[RDEXP] preview for diffraction ILL instruments

Commit:
[RDEXP] preview for reflectometry ILL instruments

Commit:
[RDEXP] preview for back scatt ILL instruments

Commit:
[RDEXP] preview for powder ILL instruments

Commit:
[RDEXP] remove unused geometry field

Commit:
[RDEXP] acqusition mode can be empty

An empty value for the acquisition mode means that the preview is valid
for any acquisition mode.

Commit:
[RDEXP] move tree widget setup to view

Commit:
[RDEXP] signal presneter when selectin changes

Commit:
[RDEXP] add getter for instrument and propagate

Commit:
[RDEXP] propagate selection to model

Commit:
[RDEXP] new view class to manipulate the widget

Commit:
[RDEXP] new class for the preview model

Commit:
[RDEXP] new class for the preview presenter

Commit:
[RDEXP] connect the new preview through MVP

Commit:
[RDEXP] connect new preview model signal

Commit:
[RDEXP] add method to delete a preview in model

Commit:
[RDEXP] new method to modify a preview

Commit:
[RDEXP] new or modify depending on the user choice

Commit:
[RDEXP] add slice viewer type

Commit:
[RDEXP] add plot 2d preview type

Commit:
[RDEXP] previews reference kept by main view

Commit:
[RDEXP] add 1d plot preview type

gvardany committed:
move the widget to where it belongs

MathieuTillet committed:
Change resizing policy to be more convenient

MathieuTillet committed:
Simpler version of the RDExp

Demo version without options. There, only the instrument viewer is
selected as a preview option, and thus most buttons can be removed

MathieuTillet committed:
[RDEXP] Relink file browser and tree widget

MathieuTillet committed:
Change plotting to be based on the dimension

MathieuTillet committed:
Change selection behaviour to accept Enter

MathieuTillet committed:
Add spectrum plot option

MathieuTillet committed:
Process data before it is shown

MathieuTillet committed:
Check for group workspaces

gvardany committed:
added a release note

MathieuTillet committed:
Change widget name to be clearer

MathieuTillet committed:
[RDEXP] Remove target option

But keep most of the underlying structure intact, just in case this
option is asked for in later versions. One never knows.

MathieuTillet committed:
[RDEXP] Remove data transformations steps

MathieuTillet committed:
[RDEXP] Sum files when multiple are selected

MathieuTillet committed:
Create and plug mechanism to determine preview

MathieuTillet committed:
Determine acquisition type by looking at data

MathieuTillet committed:
[RDEXP] Do not trigger preview when clicking a dir

MathieuTillet committed:
[RDEXP] Keep the keyboard focus on the widget

When a new preview is open, keyboard inputs will be handled by the
widget, so the user can browse without having to change the focus.

MathieuTillet committed:
[RDEXP] Catch errors when trying to accumulate

MathieuTillet committed:
[RDEXP] Clean up accumulated workspaces

Accumulated workspaces are deleted when new one are created.

MathieuTillet committed:
[RDEXP] Use Enum for acquisitions options

MathieuTillet committed:
[RDEXP] Not selecting directories anymore

MathieuTillet committed:
[RDEXP] Forbid unselecting files

MathieuTillet committed:
[RDEXP] Allow loading using the arrow keys

MathieuTillet committed:
[RDEXP] Make the window behaves as set in settings

MathieuTillet committed:
[RDEXP] Make the window behaves as set in settings

MathieuTillet committed:
[RDEXP] Add a memory manager

Add a manager so the cached data do not eat all the available memory.
It tries to smartly delete workspace if they are no needed anymore.

MathieuTillet committed:
Fix thread issue when clicking

MathieuTillet committed:
[RDEXP] Stop memory manager when closing Mantid

MathieuTillet committed:
[RDEXP] Only allow one preview at a time

Just in case this is reverted in the future, the structure in list is
kept.

MathieuTillet committed:
Fix bug in replacing ws in the instrument view

Replacing a workspace didn't change the name known to by the presenter,
which meant the window was closed when the workspace was deleted, not
the new one.

MathieuTillet committed:
[RDEXP] Unselect items when closing a widget

MathieuTillet committed:
[RDEXP] Forbid unselecting in accumulate mode

MathieuTillet committed:
[RDEXP] Allow selecting anywhere on a row

Not just on the name of the file

MathieuTillet committed:
[RDEXP] Use managed algorithm to load files

Still not showing the progress though, but it should be a step in the
good direction

MathieuTillet committed:
[RDEXP] Set starting directory for the widget

MathieuTillet committed:
[RDEXP] Ensure instrument viewer respect on top

MathieuTillet committed:
[RDEXP] Expose useThread to python

Because using a thread causes problems when the user clicks too fast, we
want not to use any. But exposing this argument to python requires it to
come before the dependencies arg (or at least it is infinitely easier to
convince sip to work this way round), so we refactor a bit the
instrument view constructor and its uses.

Also removes a closeEvent function defined twice for some reason.

MathieuTillet committed:
[RDEXP] Not use thread for the instrument viewer

Because painting the instrument actor in the background leads to
crashess if the user is too eager to see more data.

MathieuTillet committed:
[RDEXP] Set projection for some instruments

Set instrument viewer default projection for PANTHER, SHARP and IN5

MathieuTillet committed:
[RDEXP] Remove accumulate checkbox

Accumulating can now be done using the Ctrl key.

MathieuTillet committed:
[RDEXP] Expose flip view checkbox to IDF

Make it so the IDF can define the default status of the flip checkbox in
the instrument view render tab. Needed for some ILL instruments.

MathieuTillet committed:
[RDEXP] Set a waiting cursor when loading

MathieuTillet committed:
[RDEXP] Flip view by default in IN5 and SHARP

MathieuTillet committed:
[RDEXP] Forbid selection when loading

MathieuTillet committed:
[RDEXP] Downgrade to an error message

MathieuTillet committed:
[RDEXP] Remove dead code

A whole lot of it

MathieuTillet committed:
[RDEXP] Adding some documentation and cleaning

MathieuTillet committed:
[RDEXP] Fix tests

MathieuTillet committed:
[RDEXP] Change instrument view signals

Because of tests, whcih requires mocking and mocks don't like
connections.

MathieuTillet committed:
[RDEXP] Fix instrument view test

The instrument view's test checks for the number of call to
getSurface(), which is up by one because of the call to flip the view if
needed. I'm not sure of the point of this test, but at least it is up to
date

MathieuTillet committed:
[RDEXP] Not updating the iview when useless

When calling the flipping function, stop updating the iview if the
flipping aws already set to the provided value.

MathieuTillet committed:
[RDEXP] Fix IN5 IDF

MathieuTillet committed:
[RDEXP] Flip PANTHER default view

MathieuTillet committed:
[RDEXP] Lower tolerance for IN5 tests

Recreating the geometry lead to slightly different pixel positions, so
the results differs only so slightly from their reference

MathieuTillet committed:
[RDEXP] Fix instrument view test

MathieuTillet committed:
[RDEXP] Check existence before removing

Because a preview can be opened outside of the rdexp, and thus closed
with it having to remove it from the list of folowed previews

MathieuTillet committed:
[RDEXP] Hack to keep the focus on the rdexp

Since whenever a new widget is open, or whenever a matplotlib plot si
changed, the widget steals the focus and render the keyboard interaction
impossible, the focus is immediately given back to the rdexp.

This hack is used because the event generated by this focusing out is
the same as clicking, so no discrimination here, and matplotlib cannot
be convinced not to steal the focus in the first place.

MathieuTillet committed:
[RDEXP] Fix typo and merge issue

MathieuTillet committed:
[RDEXP] Allow reselecting after closing preview

MathieuTillet committed:
[RDEXP] Open file browser on last directory

MathieuTillet committed:
[RDEXP] Better management of unsupported data

Better sanitization before hand and better handling of bad data

MathieuTillet committed:
[RDEXP] Catch error when the file has no loader

MathieuTillet committed:
[RDEXP] Fix tests again

MathieuTillet committed:
[RDEXP] Open popup on error

MathieuTillet committed:
[RDEXP] Harsher condition on the iviewer

Won't try to show it if it's not possible. Also more consistent error
reporting.

MathieuTillet committed:
[RDEXP] Fix bug when clicking on dir

When clicking on a directory, the previous selection was reloaded

MathieuTillet committed:
[RDEXP] Polishing preview policy

And clairfy error messages

MathieuTillet committed:
[RDEXP] Fix accumulating behavior

MathieuTillet committed:
[RDEXP] Fix behaviour when changing preview type

MathieuTillet committed:
[RDEXP] Set default and warning for preview search

MathieuTillet committed:
[RDEXP] Change expansion policy for button

MathieuTillet committed:
[RDEXP] Warn the user when selecting an invalid directory

MathieuTillet committed:
[RDEXP] Clean conversion between types

MathieuTillet committed:
[RDEXP] Better way of handling the memory load

MathieuTillet committed:
[RDEXP] Some more docstrings

MathieuTillet committed:
[RDEXP] Call the correct module for dataobjects

MathieuTillet committed:
[RDEXP] Discover directories before new root

Before creating a new root for the qtreeview, discover the directory,
because qt is awfully slow on big ones so any improvement is good to
take

MathieuTillet committed:
[RDEXP] Fix import bc of sliceviewer refactoring

darominski committed:
[RDEXP] Fix getting memory info

darominski committed:
[RDExp] Added flip-view attribute to IDF schema

darominski committed:
[RDExp] Removed redundant attributes from ILL instruments descriptions

previously added during development of RDExp, also updated some
descriptions of ILL instruments.

darominski committed:
Revert "Add ILL/D11 acquisition mode"

This reverts commit 22100573369fa523aec49c1f359e7d47c1f1a23d.

darominski committed:
[RDExp] Removed all acquisition and attribute from Facilities.xml

as these parameters are not used and are not part of the schema.

MathieuTillet committed:
Delete file readded accidentally

MathieuTillet committed:
Fix arg swap in the file that moved

MathieuTillet committed:
[RDEXP] Fix using the incorrect closeEvent fct

MathieuTillet committed:
[RDEXP] Add type checking and complete comments

MathieuTillet committed:
[RDEXP] Update copyright date

MathieuTillet committed:
[RDEXP] Get file name in a cleaner way

MathieuTillet committed:
[RDEXP] Triggering file formatting

MathieuTillet committed:
[RDEXP] Fix spectrum never being shown

MathieuTillet committed:
[RDEXP] Use std algorithms for preview manager

Cppcheck was finding the code to unappealing to let the tests continue,
so hopefully it is more to its taste now

MathieuTillet committed:
[RDEXP] Add missing doc for parameter

darominski commented on Framework/API/CMakeLists.txt:
```diff
@@ -121,6 +121,7 @@ set(SRC_FILES
     src/ParameterReference.cpp
     src/ParameterTie.cpp
     src/PeakFunctionIntegrator.cpp
+    src/PreviewDeclarations.cpp
```
I believe this file can be safely suppressed, as the methods inside are not called anywhere, and the widget works as intended without.

darominski commented on instrument/IN5_Definition.xml:
```diff
@@ -41,664 +42,664 @@
       <axis x="0.0" y="1.0" z="0.0"/>
-      <radius val="0.0122"/>
-      <height val="0.0122406639004"/>
+      <radius val="0.012199999999999999"/>
```
I do wonder why there is so many changes to the IDFs of IN5, PANTHER, and SHARP, and most of them seem to be minimal/stylistic. Where do they stem from? Also, are they taken in the account in the IDF generation script on the mantidgeometry repository?

MathieuTillet replied:
If I recall, adding the flip parameter to the generator (not sure if I pushed that to mantid-geometry, I need to check) meant I needed to rerun the scripts and for some reason all positions got moved by a absurdly tiny distance and so the diff is huge without any relevant change

darominski commented on Framework/API/inc/MantidAPI/IPreview.h:
```diff
@@ -23,12 +23,13 @@ namespace API {
   virtual std::string technique() const = 0;
   virtual std::string facility() const = 0;
   virtual std::string name() const = 0;
+  virtual std::string acquisition() const = 0;
```
I believe the acquisition mode is not used in the current version of the rawdata explorer. It is not defined in the IDF nor IPFs, and thus remains dead code. This line can be therefore removed.

darominski commented on Framework/API/inc/MantidAPI/PreviewManager.h:
```diff
@@ -28,26 +29,31 @@ class MANTID_API_DLL PreviewManagerImpl {
-  const IPreview &getPreview(const std::string &facility, const std::string &technique,
+  std::vector<std::string> getPreviews(const std::string &facility, const std::string &technique = "",
+                                       const std::string &acquisition = "") const;
+  const IPreview &getPreview(const std::string &facility, const std::string &technique, const std::string &acquisition,
```
Here, and in the line above, the reference to the `acquisition` can be removed.

darominski commented on Framework/API/inc/MantidAPI/PreviewManager.h:
```diff
@@ -28,26 +29,31 @@ class MANTID_API_DLL PreviewManagerImpl {
     IPreview_uptr preview = std::make_unique<T>();
     const auto facility = preview->facility();
     const auto technique = preview->technique();
+    const auto acquisition = preview->acquisition();
```
To be removed.

darominski commented on Framework/API/inc/MantidAPI/PreviewManager.h:
```diff
@@ -28,26 +29,31 @@ class MANTID_API_DLL PreviewManagerImpl {
+    const auto acquisition = preview->acquisition();
     const auto name = preview->name();
-    if (checkPreview(facility, technique, name)) {
+    if (checkPreview(facility, technique, acquisition, name)) {
```
Acquisition can be removed.

darominski commented on Framework/API/inc/MantidAPI/PreviewManager.h:
```diff
@@ -28,26 +29,31 @@ class MANTID_API_DLL PreviewManagerImpl {
+    if (checkPreview(facility, technique, acquisition, name)) {
       throw std::runtime_error("Preview with the same name is already registered for the same "
-                               "facility and technique.");
+                               "facility, technique and acquisition mode.");
```
Acquisition can be removed.

darominski commented on Framework/API/inc/MantidAPI/PreviewManager.h:
```diff
@@ -28,26 +29,31 @@ class MANTID_API_DLL PreviewManagerImpl {
   bool checkFacility(const std::string &facility) const;
   bool checkTechnique(const std::string &facility, const std::string &technique) const;
-  bool checkPreview(const std::string &facility, const std::string &technique, const std::string &preview) const;
+  bool checkAcquisition(const std::string &facility, const std::string &technique,
```
This check function can be removed.

darominski commented on Framework/API/src/PreviewDeclarations.cpp:
```diff
@@ -0,0 +1,197 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
```
This entire file can be safely removed, it is not used in the current version of the RDExp.

darominski commented on Framework/API/src/PreviewManager.cpp:
```diff
@@ -4,49 +4,75 @@
-std::vector<std::string> PreviewManagerImpl::getPreviews(const std::string &facility,
-                                                         const std::string &technique) const {
+std::vector<std::string> PreviewManagerImpl::getPreviews(const std::string &facility, const std::string &technique,
+                                                         const std::string &acquisition) const {
```
The acquisition reference can be removed.

darominski commented on Framework/API/src/PreviewManager.cpp:
```diff
@@ -4,49 +4,75 @@
-      }
+  for (const auto &preview : m_previews) {
+    if (preview->facility() == facility && (technique.empty() || preview->technique() == technique) &&
+        (acquisition.empty() || preview->acquisition().empty() || preview->acquisition() == acquisition)) {
```
The checks on acquisition can be removed.

darominski commented on Framework/API/src/PreviewManager.cpp:
```diff
@@ -4,49 +4,75 @@
+  //  }
+  auto isPreview = [&](const auto &preview) {
+    return (preview->facility() == facility && preview->technique() == technique &&
+            (preview->acquisition().empty() || preview->acquisition() == acquisition) &&
```
The checks on acquisition can be removed.

darominski commented on Framework/API/src/PreviewManager.cpp:
```diff
@@ -4,49 +4,75 @@
+  // cppcheck was unhappy about this for loop, so it got replaced, but is left as a comment so it is actually readable
+  //  for (const auto &preview : m_previews) {
+  //    if (preview->facility() == facility && preview->technique() == technique &&
+  //        (preview->acquisition().empty() || preview->acquisition() == acquisition) && preview->name() ==
```
The checks on acquisition can be removed from the comment as well.

darominski commented on Framework/API/src/PreviewManager.cpp:
```diff
@@ -4,49 +4,75 @@
+  return std::any_of(std::begin(m_previews), std::end(m_previews), isTechnique);
 }
+
+bool PreviewManagerImpl::checkAcquisition(const std::string &facility, const std::string &technique,
```
The implementation of this function can be removed.

darominski commented on Framework/API/src/PreviewManager.cpp:
```diff
@@ -4,49 +4,75 @@
-  } else {
-    return false;
-  }
+                                      const std::string &acquisition, const std::string &preview_name) const {
```
The references to `acquisition` can be removed.

darominski commented on Framework/API/test/IPreviewTest.h:
```diff
@@ -28,6 +28,7 @@ class BasicPreview : public IPreview {
   std::string name() const override { return "BasicPreview"; }
   std::string facility() const override { return "TestFacility"; }
   std::string technique() const override { return "SANS"; }
+  std::string acquisition() const override { return "Mono"; }
```
This line can be safely removed.

darominski commented on Framework/API/test/IPreviewTest.h:
```diff
@@ -50,6 +51,7 @@ class IPreviewTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(preview->name(), "BasicPreview")
     TS_ASSERT_EQUALS(preview->facility(), "TestFacility")
     TS_ASSERT_EQUALS(preview->technique(), "SANS")
+    TS_ASSERT_EQUALS(preview->acquisition(), "Mono")
```
This check can be safely removed.

darominski commented on Framework/API/test/PreviewManagerTest.h:
```diff
@@ -24,6 +24,7 @@ class BasicPreview : public IPreview {
   std::string name() const override { return "BasicPreview"; }
   std::string facility() const override { return "TestFacility"; }
   std::string technique() const override { return "SANS"; }
+  std::string acquisition() const override { return "Mono"; }
```
This line can be safely removed.

darominski commented on Framework/API/test/PreviewManagerTest.h:
```diff
@@ -57,15 +58,29 @@ class PreviewManagerTest : public CxxTest::TestSuite {
     auto previews = PreviewManager::Instance().getPreviews("TestFacility", "Crystal");
     TS_ASSERT(previews.empty());
   }
+  void test_get_preview_by_facility_and_technique_and_acquisition() {
```
These tests can be reformatted to not include the acquisition, perhaps the on on `non_existent_acquisition` can be removed altogether.

MathieuTillet replied:
I believe all the tests including acquisition can be straight deleted, since removing the acquisition would make them redundant.

darominski commented on Framework/API/test/PreviewManagerTest.h:
```diff
@@ -57,15 +58,29 @@ class PreviewManagerTest : public CxxTest::TestSuite {
+  }
   void test_get_preview_by_name() {
-    auto &preview = PreviewManager::Instance().getPreview("TestFacility", "SANS", "BasicPreview");
+    auto &preview = PreviewManager::Instance().getPreview("TestFacility", "SANS", "Mono", "BasicPreview");
```
The change done in this line can be safely removed.

darominski commented on Framework/API/test/PreviewManagerTest.h:
```diff
@@ -57,15 +58,29 @@ class PreviewManagerTest : public CxxTest::TestSuite {
   }
   void test_get_preview_by_non_existent_name() {
-    TS_ASSERT_THROWS(PreviewManager::Instance().getPreview("TestFacility", "SANS", "BasicPreview2"),
+    TS_ASSERT_THROWS(PreviewManager::Instance().getPreview("TestFacility", "SANS", "Mono", "BasicPreview2"),
```
The change done in this line can be safely removed.

darominski commented on Framework/Kernel/inc/MantidKernel/InstrumentInfo.h:
```diff
@@ -55,6 +55,8 @@ class MANTID_KERNEL_DLL InstrumentInfo {
   /// Return list of techniques
   const std::set<std::string> &techniques() const;
+  /// Return list of acquisition modes
+  const std::set<std::string> &acquisitions() const;
```
This line can be safely removed.

darominski commented on Framework/Kernel/inc/MantidKernel/InstrumentInfo.h:
```diff
@@ -72,6 +74,7 @@ class MANTID_KERNEL_DLL InstrumentInfo {
 
 private:
   void fillTechniques(const Poco::XML::Element *elem);
+  void fillAcquistions(const Poco::XML::Element *elem);
```
This line can be safely removed.

darominski commented on Framework/Kernel/inc/MantidKernel/InstrumentInfo.h:
```diff
@@ -83,12 +86,13 @@ class MANTID_KERNEL_DLL InstrumentInfo {
+  ZeroPaddingMap m_zeroPadding;        ///< Run number-dependent zero padding
+  std::string m_delimiter;             ///< Delimiter between instrument name and run number
+  std::set<std::string> m_technique;   ///< List of techniques the instrument can do
+  std::set<std::string> m_acquistions; /// List of acquisition modes this instrument supports
```
This line can be safely removed.

darominski commented on Framework/Kernel/src/InstrumentInfo.cpp:
```diff
@@ -43,6 +44,7 @@ InstrumentInfo::InstrumentInfo(const FacilityInfo *f, const Poco::XML::Element *
     m_shortName = m_name;
 
   fillTechniques(elem);
+  fillAcquistions(elem);
```
This line can be safely removed.

darominski commented on Framework/Kernel/src/InstrumentInfo.cpp:
```diff
@@ -171,6 +173,8 @@ const std::vector<LiveListenerInfo> &InstrumentInfo::liveListenerInfoList() cons
 /// Return list of techniques
 const std::set<std::string> &InstrumentInfo::techniques() const { return m_technique; }
 
+const std::set<std::string> &InstrumentInfo::acquisitions() const { return m_acquistions; }
```
This line can be safely removed.

darominski commented on Framework/Kernel/src/InstrumentInfo.cpp:
```diff
@@ -244,6 +248,28 @@ void InstrumentInfo::fillTechniques(const Poco::XML::Element *elem) {
   }
 }
 
+/// Called from constructor to fill acquistions
```
This function can be safely removed.

darominski commented on Framework/PythonInterface/mantid/api/src/Exports/IPreview.cpp:
```diff
@@ -32,12 +32,14 @@ void export_IPreview() {
       .def("facility", &IPreview::facility, arg("self"), "Get the facility of the preview")
       .def("technique", &IPreview::technique, arg("self"), "Get the technique of the preview.")
       .def("type", &IPreview::type, arg("self"), "Get the type of the preview.")
+      .def("acquisition", &IPreview::acquisition, arg("self"), "Get the acquisition mode of the preview")
```
This line can be safely removed.

darominski commented on Framework/PythonInterface/mantid/api/src/Exports/PreviewManager.cpp:
```diff
@@ -20,11 +20,16 @@ PreviewManagerImpl &instance() { return PreviewManager::Instance(); }
-           "technique.")
+
+      .def("getPreview", &PreviewManagerImpl::getPreview,
+           (arg("self"), arg("facility"), arg("technique"), arg("acquisition"), arg("name")),
```
The reference to `acquisition` can be safely removed here.

darominski commented on Framework/PythonInterface/mantid/api/src/Exports/PreviewManager.cpp:
```diff
@@ -20,11 +20,16 @@ PreviewManagerImpl &instance() { return PreviewManager::Instance(); }
+
+      .def("getPreview", &PreviewManagerImpl::getPreview,
+           (arg("self"), arg("facility"), arg("technique"), arg("acquisition"), arg("name")),
+           "Get the preview by the facility, technique, acquisition mode "
```
The reference to `acquisition` can be safely removed from the comment.

darominski commented on Framework/PythonInterface/mantid/api/src/Exports/PreviewManager.cpp:
```diff
@@ -20,11 +20,16 @@ PreviewManagerImpl &instance() { return PreviewManager::Instance(); }
+           "and name.",
+           return_value_policy<reference_existing_object>())
+      .def("getPreviews", &PreviewManagerImpl::getPreviews,
+           (arg("self"), arg("facility"), arg("technique") = "", arg("acquisition") = ""),
```
The reference to `acquisition` can be safely removed here.

darominski commented on Framework/PythonInterface/mantid/api/src/Exports/PreviewManager.cpp:
```diff
@@ -20,11 +20,16 @@ PreviewManagerImpl &instance() { return PreviewManager::Instance(); }
+      .def("getPreviews", &PreviewManagerImpl::getPreviews,
+           (arg("self"), arg("facility"), arg("technique") = "", arg("acquisition") = ""),
+           "Get the names of the previews available for the facility, "
+           "technique and acquisition mode.")
```
The reference to `acquisition` can be safely removed from the comment.

darominski commented on Framework/PythonInterface/mantid/kernel/src/Exports/InstrumentInfo.cpp:
```diff
@@ -47,6 +47,9 @@ void export_InstrumentInfo() {
       .def("techniques", &InstrumentInfo::techniques, arg("self"), return_value_policy<copy_const_reference>(),
            "Return list of techniques this instrument supports")
 
+      .def("acquisitions", &InstrumentInfo::acquisitions, arg("self"), return_value_policy<copy_const_reference>(),
```
This export of `acquisition` can be safely removed.

darominski commented on instrument/Facilities.xml:
```diff
@@ -740,19 +733,15 @@
 
   <instrument name="D7">
-    <technique>Polarized diffraction</technique>
+    <technique>Polarized Diffraction</technique>
```
For consistency with diffractometers, would you agree that should these technique names be written as a single word?

MathieuTillet replied:
Seems good I guess

darominski commented on instrument/Facilities.xml:
```diff
@@ -716,15 +709,15 @@
+    <technique>PowderDiffraction</technique>
   </instrument>
 
   <instrument name="D4C">
```
Line below could be modified to `LiquidDiffraction` to follow the diffraction pattern above.

MathieuTillet replied:
That's a kind of diffraction pattern I never saw before !

MathieuTillet committed:
[RDEXP] Add PreviewView test

MathieuTillet committed:
[RDEXP] Remove acquisition parameter

It is not used and there is no plan to use it in the future anymore.

MathieuTillet committed:
[RDEXP] Remove useless file

darominski committed:
[RDEXP] Set Plus output directly to final_ws

rather than replace the LHS workspace and then rename it to final_ws.
This fixes an occasional crash that was appearing in the rename while InstrumentView
was used to display rawdata.

darominski commented on qt/python/mantidqt/mantidqt/widgets/rawdataexplorer/view.py:
```diff
@@ -0,0 +1,328 @@
+            self.sig_request_close.connect(self._widget.view.emit_close)
+            self.sig_request_close.connect(lambda: self.on_close(None, False))
+
+        if self._type == self.PLOT2D:
```
It seems that all plots have the exact same behaviour. Do you think they could be combined into one case?

darominski commented on qt/python/mantidqt/mantidqt/widgets/rawdataexplorer/view.py:
```diff
@@ -0,0 +1,328 @@
+        if self._type == self.IVIEW:
+            self._widget.replace_workspace(workspace_name)
+        elif self._type == self.SVIEW:
+            # TODO find how to replace slice viewer's current workspace
```
Is the slice view's current workspace not replaced at all then? Could we improve on that?

MathieuTillet replied:
It certainly looks like that, though without test data it is a bit hard to be sure of the actual behavior. In any case putting _something_ in this clause would be a good first step

MathieuTillet replied:
Ok, after a look at the slice viewer code:
There is no method for replacing the workspace currently shown BUT if the workspace shown is replaced in the ADS while keeping the same name, the slice viewer adapts.
That can be an ugly hack to make this work if the need arises in the future (for now nothing actually needs that).

darominski replied:
Since nothing currently uses that, it could be either removed altogether or a warning raised when a user tries to visualise something in this mode. In the latter case, also a comment could be added here in the code.

darominski commented on qt/python/mantidqt/mantidqt/widgets/rawdataexplorer/test/test_view.py:
```diff
@@ -0,0 +1,244 @@
+        self.preview_view.on_close.assert_called_once()
+        self.preview_view._widget.container.emit_close.assert_called_once()
+
+        # TODO check also for closing connection ?
```
If possible, why not

MathieuTillet committed:
[RDEXP] Add test for the view

Very incomplete and unsatisfactory as it is, but that is all I could
achieve without a proper file system

MathieuTillet committed:
[RDEXP] Fix instrument test by calling correct arg

MathieuTillet committed:
[RDEXP] Use NiceMock instead of StrictMock

At least in the instrument widget test, since that's what is being
changed

Per explicit documentation recommandation, "use nice mocks most of the
time [...] use strict mocks only as the last resort"

MathieuTillet commented on qt/python/mantidqt/mantidqt/widgets/rawdataexplorer/model.py:
```diff
@@ -0,0 +1,305 @@
+            error_reporting("Error, could not load file:\n {0}".format(e))
+            return False
+
+        if not load_alg.isExecuted():
```
I don't think it is. L167 checks that the algorithm hasn't thrown an error, but I can imagine cases where no error is thrown yet the algorithm failed to execute. Is it possible for any loader ? I don't know, but I don't know otherwise either.

MathieuTillet commented on qt/python/mantidqt/mantidqt/widgets/rawdataexplorer/RawDataExplorerFileTree.py:
```diff
@@ -0,0 +1,153 @@
+        file_model = QFileSystemModel()
+        file_model.setNameFilters(self._FILE_SYSTEM_FILTERS)
+        file_model.setNameFilterDisables(0)
+        file_model.setRootPath("/")
```
Probably, on it.

MathieuTillet replied:
Well that was easy

MathieuTillet commented on qt/scientific_interfaces/Direct/ALFInstrumentView.cpp:
```diff
@@ -39,8 +39,8 @@ ALFInstrumentView::ALFInstrumentView(QWidget *parent)
   m_instrumentWidget =
-      new MantidWidgets::InstrumentWidget(QString::fromStdString(fileName), nullptr, true, true, 0.0, 0.0, true,
-                                          MantidWidgets::InstrumentWidget::Dependencies(), false);
+      new MantidWidgets::InstrumentWidget(QString::fromStdString(fileName), nullptr, true, true, 0.0, 0.0, true, false,
```
I don't see how it could at first glance, this is just swapping around two parameters, but if you see any potential issue let me know.

MathieuTillet committed:
[RDEXP] Remove root dir of file tree set to root

This meant we had to discover another directory and is not actually
needed in the end

MathieuTillet committed:
[RDEXP] Factorize logic for showing mpl widgets

MathieuTillet committed:
[RDEXP] Better way of determining the acquisition

Bit more robust this way, at least for D20 and D7 data which the
relevant ones.

MathieuTillet committed:
[RDEXP] Simplify accumulate naming scheme

MathieuTillet committed:
[RDEXP] Attempt at making paths less system-dependent

MathieuTillet committed:
[RDEXP] Remove old release notes badly merged

MathieuTillet committed:
[RDEXP] Standardize technique names

MathieuTillet commented on qt/applications/workbench/workbench/plugins/rawdataexplorerwidget.py:
```diff
@@ -0,0 +1,43 @@
+    def get_plugin_title(self):
+        return "Raw Data"
+
+    def readSettings(self, _):
```
Yes, else we use the base class implementation which just raises a NotImplementedError.

MathieuTillet committed:
[RDEXP] Add relevant documentation

darominski committed:
Set using thread by InstrumentWidget to false by default

This fixes found issue that unsupported workspace types, e.g. moving workspaces,
were still attempting to be drawn in InstrumentView widget and resulting in an unhandled
exception.

MathieuTillet committed:
[RDEXP] Attempt at fixing path issues on windows

MathieuTillet committed:
[RDEXP] Attempt at fixing root path issues

MathieuTillet committed:
[RDEXP] Add test for PreviewPresenter and presenter

MathieuTillet committed:
[RDEXP] Add tests for the model and PreviewModel

The most work intensive functions are still missing though, and they are
the most important to properly test

MathieuTillet committed:
[RDEXP] Remove useless has_calls for simpler called_once_with

darominski committed:
[RDEXP] Minor corrections in the doc

darominski committed:
[RDEXP] Test naming scheme consistency correction

it seems that without widget name in the name of the test file it was not possible
to call rawdataexplorer tests specifically as the root of the test name is not in the
directory of the rawdataexplorer but mantid_qt5.

darominski committed:
[RDEXP] Made path OS-independent

darominski committed:
[RDEXP] Unit test for setting working directory

darominski committed:
[RDEXP] Added tearDown class

darominski committed:
[RDEXP] Modify preview unit test

darominski committed:
[RDEXP] Unit test for choosing preview

darominski committed:
[RDEXP] Removed redundant TODOs and tests

darominski committed:
[RDEXP] Unit test for select_last_clicked

darominski committed:
[RDEXP] Added release note

darominski commented:
I finished writing up the Qt tests. The remaining discussion point is a test requiring mocking of a file system, which would add a new dependency, and then still the possible gain would be uncertain. Any suggestions on how to best implement `test_on_item_selected` from the view testing are very welcome.

darominski committed:
Merge remote-tracking branch 'origin/main' into 29249_raw_data_explorer_widget

darominski committed:
Merge remote-tracking branch 'origin/main' into 29249_raw_data_explorer_widget

darominski committed:
Merge remote-tracking branch 'origin/main' into 29249_raw_data_explorer_widget

darominski committed:
Merge remote-tracking branch 'origin/main' into 29249_raw_data_explorer_widget

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

darominski committed:
trigger build commit

darominski commented:
@gemmaguest @martyngigg Could we get a second review by someone on your side after the release?

darominski committed:
Merge branch 'main' into 29249_raw_data_explorer_widget

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

darominski committed:
Merge branch 'main' into 29249_raw_data_explorer_widget

darominski committed:
Pleasing black formatting

darominski committed:
[RDEXP] Moved release notes to v6.7

darominski reviewed (dismissed):
Already reviewed, waiting for the second round.

DannyHindson commented:
Just reviewing a few PRs as gatekeeper. Please can I check where the new widget should be displayed? For me it shows up here:

![image](https://user-images.githubusercontent.com/56295817/222404455-27cc7671-284f-41cd-8735-0b70fbf6e9c5.png)

From the description it should perhaps be an extra tab behind the workspace widget? I'm wondering if this is similar to a conversation we have about the workspace calculator (see this PR https://github.com/mantidproject/mantid/pull/31919)

perenon commented:
Hi Danny, thanks for your comment. On ILL side, I take back the responsibility of this PR.

The RawDataExplorer widget can be put behind the workspace one if you want to. As a user, that would not be my choice since I my workspace widget size is generally small. Whatsoever, the position is persistent (so is stored somehow when Mantid closes).

So what would you advise ? To change the default location of this widget ?

perenon commented:
By the way, thanks for handling that as a GK. However, since this PR is large and has a huge impact, it would be nice of someone at ISIS could have a deep look on the code.

You are now assign as a reviewer, please tell me if you so not have the bandwidth to do it.

jhaigh0 commented:
Haven't looked at the code yet, but in the interest of getting as many people as possible to look at this, I've been playing around with it this morning. I think the raw data explorer is a really nice addition to Mantid, and I have some feedback on its usability.

Also, just to note, I can't access the design document link since I don't have an ILL account.

## 1. File explorer interface

My main feedback is that I'd like the RDE to behave more like a typical file browser (i.e the Windows file explorer).

- 1a. At the moment, if you left or right click on any file, it begins the process of loading it. I would prefer that you have to double-click on a file or right click -> load.
- 1b. Following on from 1a, I think a right click menu could have a lot of uses. For example, I think there should be an option to load in data without the instrument viewer or a plot opening, maybe this is even the default.
- 1c. Also following on from 1a, I think that being able to left-click + shift / control on multiple files before double-clicking (or right-clicking) to load them into a group (or individually etc. there would be options in the right-click menu) would be a more intuitive way of loading multiple files.
- 1d. The current file explorer has three columns for name, size, and date modified; I think it would be nice if you could sort by each of these (ascending and descending), like a typical OS file browser.
- 1e. It would be a nice feature to include tab completion in the file path text box; definitely not necessary though.

## 2. Small bugs

- 2a. Loading peak data (for example `UnitTest/SingleCrystalPeakTable.nxs` results in an error box showing which says it cannot open this type of data. It does still load in the data though so I think the message box is maybe not necessary since there is a log in the console as well. This might be an area where my suggestion in 1b could be a useful, having a right-click menu choose to either just loading the data or trying to display it again. If the user chose this second option, I think the message box makes sense.
- 2b. When I first opened Mantid, the window extended a little off the bottom of my screen, but was fixed upon full-screening the program.
- 2c. If you put some data in a directory with a non-ascii name, e.g. `../über_folder/file.nxs`, and try to load it into Mantid the normal way, It is well handled by showing a message box saying there is an invalid property. If you try to load this data using the RDE, for me at least, it brought up the error reporter. I think there should be a way to handle this better. (this is a very niche issue though).

## 3. Other Behaviour

- 3a. When loading multiple files using `ctrl`, if you clicked on files `a`, `b`, `c` (in that order) you end up loading 6 workspaces (`a`, `b`, `c`, `a-b`, `a-c`). I think this is a little annoying if you just wanted to load `a-c`. My suggestion in 1c may help with this.
- 3b. Loading a group workspace (e.g. `GEM63437_Focused` from the training course data) bring up a plot of the different counts in each individual workspace. I'm not sure how useful this is.


Hopefully some of these are helpful suggestions. Once again, I can't see the design doc, so apologies if any of these suggestions are already noted down. I might continue to update this comment if I find anything else worth mentioning, but overall this is a great feature.

perenon commented:
Hi Jonathan, thanks a lot for your detailed and useful answer.
The design doc should be attached to this comment.
[RDExp.pdf](https://github.com/mantidproject/mantid/files/11321026/RDExp.pdf)

We will need some time to solve the bugs you found out. We will come back on this as soon as we can.

Considering the behavior of the interface, we wanted to be able to quickly pass from one file to an other. That explains the "one-click" opening (you can also load the file by browsing using the keyboard).
Maybe we should add a button to enable the user to choose the behavior of the interface (?).

jhaigh0 commented:
> Hi Jonathan, thanks a lot for your detailed and useful answer. The design doc should be attached to this comment. [RDExp.pdf](https://github.com/mantidproject/mantid/files/11321026/RDExp.pdf)
> 
> We will need some time to solve the bugs you found out. We will come back on this as soon as we can.
> 
> Considering the behavior of the interface, we wanted to be able to quickly pass from one file to an other. That explains the "one-click" opening (you can also load the file by browsing using the keyboard). Maybe we should add a way to enable the user to choose the behavior of the interface (?).

Thank you for the link, I'll give it a read today.

Yes, I think I didn't quite understand the intention of moving quickly from file to file at first. I think it works quite well for this.

What I'm thinking is that some users may use the RDE as a way to quickly load in files since it is more convenient than going through the Load dialogue. In this case, I think it would be useful to have a 'load only' setting. In particular, I think new users will see the RDE and may just think of it as a file browser.

It's up to you though, I think it meets its intended behaviour well.

One other (very small) thing I've thought about is that, when you click a file, the mouse cursor changes to a spinning wheel; I think it should stay as a regular cursor (since it is loading on a separate thread). When control-clicking on multiple files, the spinning cursor makes it feel like you need to wait before each is loaded before you can click on the next, which is not the case.

sf1919 commented:
@perenon I have cancelled the tests that were rerunning on this PR as there are some merge conflicts

eurydice76 committed:
Merge branch 'main' into 29249_raw_data_explorer_widget

eurydice76 committed:
Merging with main

eurydice76 committed:
Added missing include

eurydice76 committed:
Replaced message boxes by logger warnings

perenon commented:
Up

perenon commented:
(Re-)Up

perenon commented:
(Yet another) up
